### PR TITLE
Fix #7494: check type bounds during java joint compilation

### LINKED ISSUE #7494: The inferred AppliedTypes from Java are not checked
We could construct `AppliedType`s from Java whose arguments do not conform to type bound. When the type is used in Dotty, it is not checked. 

## minimized code

`S.scala`
```scala
class A

class D[T >: A](v: T) {
  def getV(): T = v // ensure T is correctly inferred
}

object S {
  // J.getDS() : D[String] is inferred but not checked
  val dv: String = J.getDS().getV()
}
```

`J.java`
```java
public class J {
  // for java, D is D<T extends Object>
  public static D<String> getDS() {
    return new D<String>("DS");
  }
}
```

## expectation

I expected to see a type error at `J.getDS()` in `S.scala`, since `String` does not conform to lower bound `A`. However, this test passed. I also test the same code in scalac 2.13, and it also compiled.

### PR DESCRIPTION
**Pull request by TheElectronWill**
As @smarter pointed out on gitter, checking every `Select` of a java-defined symbol would probably be very costly (and still not provide a bullet-proof solution). With this PR, java compilation units are discarded just before Pickler and thus go through additional checks.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala [hunk 1]
```diff
@@ -85,6 +92,69 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
...
+        case tp: PolyType if args.exists(isNamedArg) =>
+          val (namedArgs, otherArgs) = args.partition(isNamedArg)
+          val args1 = reorderArgs(tp.paramNames, namedArgs.asInstanceOf[List[NamedArg]], otherArgs)
+          TypeApply(tycon, args1).withSpan(tree.span).withType(tree.tpe)
+        case _ =>
+          tree
+  end normalizeTypeArgs
+
+  /** Scan a tree and check all the AppliedTypeTrees in it. */
+  def checkAppliedTypes(tree: Tree)(using Context): Unit = tree match
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/PostTyper.scala:130]: I think this could be done using a TreeTraverser

Comment (TheElectronWill) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala]: That simplifies things, thanks!

**Comment by smarter** [comment]
There's one failure in the community build:

> 2020-07-15T22:58:10.6062064Z utest.jvm[0.26.0-bin-SNAPSHOT].compile java.lang.AssertionError: assertion failed: The startSource can be called only once per source file: /__w/dotty/dotty/community-build/community-projects/utest/utest/jvm/src-0/utest/EnableReflectiveInstantiation.java

Which is probably related to the sbt.ExtractDependencies now being run on java sources. I didn't realize we had all this stuff before PostTyper which should not be run on Java units indeed. I think the simplest way to achieve that might be to go back to dropping these units after Frontend, and just have Frontend take care of running the checks on the Java units before it drops them.

File: compiler/src/dotty/tools/dotc/typer/FrontEnd.scala [hunk 2]
```diff
@@ -80,6 +80,8 @@ class FrontEnd extends Phase {
         typr.println("typed: " + unit.source)
         record("retained untyped trees", unit.untpdTree.treeSize)
         record("retained typed trees after typer", unit.tpdTree.treeSize)
+        if unit.isJava then
+          JavaChecks.check(unit.tpdTree)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/FrontEnd.scala:84]: I think we should run these checks in `runOn` on all Java units at once after `unitContexts.foreach(typeCheck(using _))` for the same reason we normally run them in PostTyper: doing these checks in the middle of typechecking could lead to cycles

Comment (TheElectronWill) [reply compiler/src/dotty/tools/dotc/typer/FrontEnd.scala:84]: done!

**Comment by TheElectronWill** [comment]
I agree, it's clearer that way. I've put some explanations in the commit message.

**Pull request merged**